File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel.Tests/Utilities/EventListTests.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Tests.Utilities
{
    using Microsoft.CommonDataModel.ObjectModel.Cdm;
    using Microsoft.CommonDataModel.ObjectModel.Enums;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using Microsoft.VisualStudio.TestTools.UnitTesting;
    using System.Threading.Tasks;

    /// <summary>
    /// Collection of tests to verify event recorder behavior.
    /// </summary>
    [TestClass]
    public class EventListTests
    {
        /// <summary>
        /// The path between TestDataPath and TestName.
        /// </summary>
        private readonly string testsSubpath = "Utilities";

        /// <summary>
        /// Dummy value used for correlation ID testing.
        /// </summary>
        private readonly string DummyCorrelationId = "12345";

        /// <summary>
        /// Declare a blackhole callback. We're focusing on event recorder, don't care about output going to the standard log stream.
        /// </summary>
        private readonly EventCallback eventCallback = new EventCallback
        {
            Invoke = (level, message) =>
            {
                // NOOP
            }
        };

        /// <summary>
        /// Tests several use cases where no nesting of recording functions happens.
        /// </summary>
        /// <returns></returns>
        [TestMethod]
        public async Task TestWithoutNesting()
        {
            CdmCorpusDefinition corpus = TestHelper.GetLocalCorpus(testsSubpath, "TestEventList");
            corpus.SetEventCallback(eventCallback, CdmStatusLevel.Warning, DummyCorrelationId);

            // Test fetching an object from invalid namespace results in at least one error message in the recorder
            _ = await corpus.FetchObjectAsync<CdmDocumentDefinition>("foo:/bar");
            TestBasicLogsState(corpus);
            TestHelper.AssertCdmLogCodeEquality(corpus, CdmLogCode.ErrStorageNamespaceNotRegistered, true);

            // Test fetching a good object, this should leave event recorder empty
            _ = await corpus.FetchObjectAsync<CdmDocumentDefinition>("local:/default.manifest.cdm.json");
            TestNoLogsState(corpus);

            // Test saving a manifest to invalid namespace results in at least one error message in the recorder
            var manifest = corpus.MakeObject<CdmManifestDefinition>(CdmObjectType.ManifestDef, "dummy");
            await manifest.SaveAsAsync("foo:/bar", true);
            TestBasicLogsState(corpus);
            TestHelper.AssertCdmLogCodeEquality(corpus, CdmLogCode.ErrValdnMissingDoc, true);

            // Test resolving a manifest not added to a folder, this should yield at least one error message in the recorder
            await manifest.CreateResolvedManifestAsync("new dummy", null);
            TestBasicLogsState(corpus);
            TestHelper.AssertCdmLogCodeEquality(corpus, CdmLogCode.ErrResolveManifestFailed, true);

            // Test resolving an entity without WRT doc, this should yield at least one error message in the recorder
            var entity2 = corpus.MakeObject<CdmEntityDefinition>(CdmObjectType.EntityDef, "MyEntity2");
            await entity2.CreateResolvedEntityAsync("MyEntity2-Resolved");
            TestBasicLogsState(corpus);
            TestHelper.AssertCdmLogCodeEquality(corpus, CdmLogCode.ErrDocWrtDocNotfound, true);

            // Test invoking FileStatusCheckAsync on the manifest, this should yield at least one error message in the recorder
            await manifest.FileStatusCheckAsync();
            TestBasicLogsState(corpus);
            TestHelper.AssertCdmLogCodeEquality(corpus, CdmLogCode.ErrStorageNullNamespace, true);

            // Repeat the same test but with status level 'None', no events should be recorded
            corpus.Ctx.ReportAtLevel = CdmStatusLevel.None;
            await entity2.CreateResolvedEntityAsync("MyEntity2-Resolved");
            TestNoLogsState(corpus);

            // Test checking file status on a data partition
            // We're at log level 'Progress', so we get the EnterScope/LeaveScope messages too
            corpus.Ctx.ReportAtLevel = CdmStatusLevel.Progress;
            var part = corpus.MakeObject<CdmDataPartitionDefinition>(CdmObjectType.DataPartitionDef, "part");
            await part.FileStatusCheckAsync();
            TestBasicLogsState(corpus);
            TestHelper.AssertCdmLogCodeEquality(corpus, CdmLogCode.ErrPathNullObjectPath, true);

            // Test checking file status on a data partition pattern
            var refDoc = corpus.MakeObject<CdmDocumentDefinition>(CdmObjectType.DocumentDef, "RefEntDoc");
            var partPattern = corpus.MakeObject<CdmDataPartitionPatternDefinition>(CdmObjectType.DataPartitionPatternDef, "partPattern");
            partPattern.InDocument = refDoc;
            await partPattern.FileStatusCheckAsync();
            TestBasicLogsState(corpus);
            TestHelper.AssertCdmLogCodeEquality(corpus, CdmLogCode.ErrStorageNullCorpusPath, true);

            // Test calculating relationships - no errors/warnings
            await corpus.CalculateEntityGraphAsync(manifest);
            TestBasicLogsState(corpus);

            // Test populating relationships in manifest - no errors/warnings
            await manifest.PopulateManifestRelationshipsAsync();
            TestBasicLogsState(corpus);

            // Test filtering code logic 
            corpus.Ctx.SuppressedLogCodes.Add(CdmLogCode.ErrPathNullObjectPath);
            var part2 = corpus.MakeObject<CdmDataPartitionDefinition>(CdmObjectType.DataPartitionDef, "part2");
            await part2.FileStatusCheckAsync();

            TestHelper.AssertCdmLogCodeEquality(corpus, CdmLogCode.ErrPathNullObjectPath, false);
        }

        /// <summary>
        /// Tests a use case where recording nesting happens, such as CreateResolvedManifestAsync making calls to CreateResolvedEntityAsync
        /// </summary>
        /// <returns></returns>
        [TestMethod]
        public async Task TestWithNesting()
        {
            CdmCorpusDefinition corpus = TestHelper.GetLocalCorpus(testsSubpath, "TestEventList");
            corpus.SetEventCallback(eventCallback, CdmStatusLevel.Warning, DummyCorrelationId);

            var manifest = corpus.MakeObject<CdmManifestDefinition>(CdmObjectType.ManifestDef, "dummy");

            // Test resolving a manifest (added to a folder) with an entity in it, this should collect messages from
            // CreateResolvedManifestAsync and CreateResolvedEntityAsync functions
            corpus.Storage.FetchRootFolder("local").Documents.Add(manifest);
            var entity1 = corpus.MakeObject<CdmEntityDefinition>(CdmObjectType.EntityDef, "MyEntity1");

            var someAttrib1 = corpus.MakeObject<CdmTypeAttributeDefinition>(CdmObjectType.TypeAttributeDef, "someAttrib1", false);
            someAttrib1.DataType = corpus.MakeRef<CdmDataTypeReference>(CdmObjectType.DataTypeRef, "entityId", true);
            entity1.Attributes.Add(someAttrib1);

            var entity1Doc = corpus.MakeObject<CdmDocumentDefinition>(CdmObjectType.DocumentDef, "MyEntity1.cdm.json");
            entity1Doc.Definitions.Add(entity1);
            corpus.Storage.FetchRootFolder("local").Documents.Add(entity1Doc);

            manifest.Entities.Add(entity1);
            await manifest.CreateResolvedManifestAsync("new dummy 2", null);

            TestBasicLogsState(corpus);
            TestHelper.AssertCdmLogCodeEquality(corpus, CdmLogCode.ErrResolveReferenceFailure, true);

            // Keep for debugging
            //corpus.Ctx.Events.ForEach(logEntry => {
            //    logEntry.ToList().ForEach(logEntryPair => Console.WriteLine($"{logEntryPair.Key}={logEntryPair.Value}"));
            //    Console.WriteLine();
            //});
        }

        /// <summary>
        /// Tests events generated in StorageManager APIs
        /// </summary>
        /// <returns></returns>
        [TestMethod]
        public void TestStorageManagerEvents()
        {
            CdmCorpusDefinition corpus = TestHelper.GetLocalCorpus(testsSubpath, "TestEventList");
            corpus.SetEventCallback(eventCallback, CdmStatusLevel.Info, DummyCorrelationId);

            corpus.Storage.Mount("dummy", null);
            TestBasicLogsState(corpus);
            TestHelper.AssertCdmLogCodeEquality(corpus, CdmLogCode.ErrStorageNullAdapter, true);

            corpus.Storage.Unmount("nothing");
            TestBasicLogsState(corpus);
            TestHelper.AssertCdmLogCodeEquality(corpus, CdmLogCode.WarnStorageRemoveAdapterFailed, true);

            // No errors/warnings expected here
            corpus.Storage.FetchRootFolder(null);
            TestBasicLogsState(corpus);
            TestHelper.AssertCdmLogCodeEquality(corpus, CdmLogCode.ErrStorageNullNamespace, true);

            corpus.Storage.AdapterPathToCorpusPath("Test");
            TestBasicLogsState(corpus);
            TestHelper.AssertCdmLogCodeEquality(corpus, CdmLogCode.ErrStorageInvalidAdapterPath, true);

            corpus.Storage.CorpusPathToAdapterPath("unknown:/Test");
            TestBasicLogsState(corpus);
            TestHelper.AssertCdmLogCodeEquality(corpus, CdmLogCode.ErrStorageAdapterNotFound, true);
            TestHelper.AssertCdmLogCodeEquality(corpus, CdmLogCode.ErrStorageNamespaceNotRegistered, true);

            corpus.Storage.CreateAbsoluteCorpusPath(null);
            TestBasicLogsState(corpus);
            TestHelper.AssertCdmLogCodeEquality(corpus, CdmLogCode.ErrPathNullObjectPath, true);

            corpus.Storage.CreateRelativeCorpusPath(null);
            TestBasicLogsState(corpus);
            TestHelper.AssertCdmLogCodeEquality(corpus, CdmLogCode.ErrPathNullObjectPath, true);
        }

        /// <summary>
        /// Test logging of API scope entry/exit.
        /// </summary>
        /// <returns></returns>
        [TestMethod]
        public async Task TestScoping()
        {
            CdmCorpusDefinition corpus = TestHelper.GetLocalCorpus(testsSubpath, "TestEventList");
            // For scoping test we need to use log level Info
            corpus.SetEventCallback(eventCallback, CdmStatusLevel.Info, DummyCorrelationId);

            // Test fetching a good object, this should result event recorder empty
            _ = await corpus.FetchObjectAsync<CdmDocumentDefinition>("local:/default.manifest.cdm.json");
            Assert.IsFalse(corpus.Ctx.Events.IsRecording, "Recording should be disabled at the end of API call");
            Assert.IsTrue(corpus.Ctx.Events.Count > 2, "There should have been at least 2 (debug) events recorded when fetching object with correct path");

            TestBasicLogsState(corpus);
            
            // Verify method on entry/exit event
            Assert.IsTrue(corpus.Ctx.Events[0]["method"] == "FetchObjectAsync<CdmDocumentDefinition>", "The first recorded event message should have specified scope method of 'FetchObjectAsync'");
            Assert.IsTrue(corpus.Ctx.Events[corpus.Ctx.Events.Count - 1]["method"] == "FetchObjectAsync<CdmDocumentDefinition>",
                "The last recorded event message should have specified scope method of 'FetchObjectAsync'");

            // Keep for debugging
            //corpus.Ctx.Events.ForEach(logEntry => {
            //    logEntry.ToList().ForEach(logEntryPair => Console.WriteLine($"{logEntryPair.Key}={logEntryPair.Value}"));
            //    Console.WriteLine();
            // });
        }

        /// <summary>
        /// Helper function to test that recording is stopped and no logs are recorded.
        /// </summary>
        /// <param name="corpus"></param>
        private void TestNoLogsState(CdmCorpusDefinition corpus)
        {
            TestBasicLogsState(corpus, true);
        }

        /// <summary>
        /// Helper function to check for event list state and presence of scope enter/leave logs.
        /// </summary>
        /// <param name="corpus"></param>
        /// <param name="expectNoLogs">If true, tests that recording is stopped and there are no logs in EventList. If false,
        /// tests that there are multiple entries and log enter/exit events were logged.</param>
        private void TestBasicLogsState(CdmCorpusDefinition corpus, bool expectNoLogs = false)
        {
            Assert.IsNotNull(corpus.Ctx.Events, "Events list should not be null");
            Assert.IsFalse(corpus.Ctx.Events.IsRecording, "Recording should be disabled at the end of API call");

            if (expectNoLogs)
            {
                Assert.IsTrue(corpus.Ctx.Events.Count == 0, "There should have been no events recorded when fetching object with correct path");
            }
            else
            {
                Assert.IsTrue(corpus.Ctx.Events.Count > 0, "There should have been at least one event recorded");
                Assert.IsTrue(corpus.Ctx.Events[0].ContainsKey("timestamp"), "The recorded event should have had a timestamp key");
                Assert.IsTrue(corpus.Ctx.Events[0]["cid"] == DummyCorrelationId, "The recorded event should have had a correlationId key with the dummy value");

                if (corpus.Ctx.ReportAtLevel == CdmStatusLevel.Progress)
                {
                    Assert.IsTrue(corpus.Ctx.Events[0]["message"] == "Entering scope",
                        "The first recorded event message should have specified that new scope was entered");
                    Assert.IsTrue(corpus.Ctx.Events[corpus.Ctx.Events.Count - 1]["message"].StartsWith("Leaving scope"),
                         "The last recorded event message should have specified that new scope was exited");
                }
            }
        }
    }
}

File: objectModel/Java/objectmodel/src/test/java/com/microsoft/commondatamodel/objectmodel/utilities/EventListTests.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

package com.microsoft.commondatamodel.objectmodel.utilities;

import com.microsoft.commondatamodel.objectmodel.TestHelper;
import com.microsoft.commondatamodel.objectmodel.cdm.*;
import com.microsoft.commondatamodel.objectmodel.enums.CdmLogCode;
import com.microsoft.commondatamodel.objectmodel.enums.CdmObjectType;
import com.microsoft.commondatamodel.objectmodel.enums.CdmStatusLevel;
import org.testng.annotations.Test;

import static org.testng.Assert.*;

public class EventListTests {
    /** The path between TestDataPath and TestName. */
    private static final String TESTS_SUBPATH = "Utilities";

    /** Dummy value used for correlation ID testing. */
    private static final String DUMMY_CORRELATION_ID = "12345";

    /** Declare a blackhole callback. We're focusing on event recorder, don't care about output going to the standard log stream. */
    private final EventCallback eventCallback = (level, message) -> {
        // NOOP
    };

    /**
     * Tests several use cases where no nesting of recording functions happens.
     * @throws InterruptedException
     */
    @Test
    public void testWithoutNesting() throws InterruptedException {
        CdmCorpusDefinition corpus = TestHelper.getLocalCorpus(TESTS_SUBPATH, "TestEventList");
        corpus.setEventCallback(eventCallback, CdmStatusLevel.Warning);
        corpus.getCtx().setCorrelationId(DUMMY_CORRELATION_ID);

        // Test fetching an object from invalid namespace results in at least one error message in the recorder
        corpus.fetchObjectAsync("foo:/bar").join();
        assertNotNull(corpus.getCtx().getEvents(), "Ctx.events should not be null");
        testBasicLogsState(corpus);
        TestHelper.assertCdmLogCodeEquality(corpus, CdmLogCode.ErrStorageNamespaceNotRegistered);

        // Test fetching a good object, this should leave event recorder empty
        corpus.fetchObjectAsync("local:/default.manifest.cdm.json").join();
        testNoLogsState(corpus);

        // Test saving a manifest to invalid namespace results in at least one error message in the recorder
        CdmManifestDefinition manifest = corpus.makeObject(CdmObjectType.ManifestDef, "dummy");
        manifest.saveAsAsync("foo:/bar", true).join();
        testBasicLogsState(corpus);
        TestHelper.assertCdmLogCodeEquality(corpus, CdmLogCode.ErrValdnMissingDoc);

        // Test resolving a manifest not added to a folder, this should yield at least one error message in the recorder
        manifest.createResolvedManifestAsync("new dummy", null).join();
        testBasicLogsState(corpus);
        TestHelper.assertCdmLogCodeEquality(corpus, CdmLogCode.ErrResolveManifestFailed);

        // Test resolving an entity without WRT doc, this should yield at least one error message in the recorder
        CdmEntityDefinition entity2 = corpus.makeObject(CdmObjectType.EntityDef, "MyEntity2");
        entity2.createResolvedEntityAsync("MyEntity2-Resolved").join();
        testBasicLogsState(corpus);
        TestHelper.assertCdmLogCodeEquality(corpus, CdmLogCode.ErrDocWrtDocNotfound);

        // Test invoking FileStatusCheckAsync on the manifest, this should yield at least one error message in the recorder
        manifest.fileStatusCheckAsync().join();
        testBasicLogsState(corpus);
        TestHelper.assertCdmLogCodeEquality(corpus, CdmLogCode.ErrStorageNullNamespace);

        // Repeat the same test but with status level 'None', no events should be recorded
        corpus.getCtx().setReportAtLevel(CdmStatusLevel.None);
        entity2.createResolvedEntityAsync("MyEntity2-Resolved").join();
        testNoLogsState(corpus);

        // Test checking file status on a data partition
        // We're at log level 'Progress', so we get the EnterScope/LeaveScope messages too
        corpus.getCtx().setReportAtLevel(CdmStatusLevel.Progress);
        CdmDataPartitionDefinition part = corpus.makeObject(CdmObjectType.DataPartitionDef, "part");
        part.fileStatusCheckAsync().join();
        testBasicLogsState(corpus);
        TestHelper.assertCdmLogCodeEquality(corpus, CdmLogCode.ErrPathNullObjectPath);

        // Test checking file status on a data partition pattern
        CdmDocumentDefinition refDoc = corpus.makeObject(CdmObjectType.DocumentDef, "RefEntDoc");
        CdmDataPartitionPatternDefinition partPattern = corpus.makeObject(CdmObjectType.DataPartitionPatternDef, "partPattern");
        partPattern.setInDocument(refDoc);
        partPattern.fileStatusCheckAsync().join();
        testBasicLogsState(corpus);
        TestHelper.assertCdmLogCodeEquality(corpus, CdmLogCode.ErrStorageNullNamespace);
        TestHelper.assertCdmLogCodeEquality(corpus, CdmLogCode.ErrDocAdapterNotFound);

        // Test calculating relationships - no errors/warnings
        corpus.calculateEntityGraphAsync(manifest).join();
        testBasicLogsState(corpus);

        // Test populating relationships in manifest - no errors/warnings
        manifest.populateManifestRelationshipsAsync().join();
        testBasicLogsState(corpus);
    }

    /**
     * Tests a use case where recording nesting happens, such as CreateResolvedManifestAsync making calls to CreateResolvedEntityAsync
     * @throws InterruptedException
     */
    @Test
    public void testWithNesting() throws InterruptedException {
        CdmCorpusDefinition corpus = TestHelper.getLocalCorpus(TESTS_SUBPATH, "TestEventList");
        corpus.setEventCallback(eventCallback, CdmStatusLevel.Warning);
        corpus.getCtx().setCorrelationId(DUMMY_CORRELATION_ID);

        CdmManifestDefinition manifest = corpus.makeObject(CdmObjectType.ManifestDef, "dummy");

        // Test resolving a manifest (added to a folder) with an entity in it, this should collect messages from
        // CreateResolvedManifestAsync and CreateResolvedEntityAsync functions
        corpus.getStorage().fetchRootFolder("local").getDocuments().add(manifest);
        CdmEntityDefinition entity1 = corpus.makeObject(CdmObjectType.EntityDef, "MyEntity1");

        CdmTypeAttributeDefinition someAttrib1 = corpus.makeObject(CdmObjectType.TypeAttributeDef, "someAttrib1", false);
        someAttrib1.setDataType(corpus.makeRef(CdmObjectType.DataTypeRef, "entityId", true));
        entity1.getAttributes().add(someAttrib1);

        CdmDocumentDefinition entity1Doc = corpus.makeObject(CdmObjectType.DocumentDef, "MyEntity1.cdm.json");
        entity1Doc.getDefinitions().add(entity1);
        corpus.getStorage().fetchRootFolder("local").getDocuments().add(entity1Doc);

        manifest.getEntities().add(entity1);
        manifest.createResolvedManifestAsync("new dummy 2", null).join();

        testBasicLogsState(corpus);
        TestHelper.assertCdmLogCodeEquality(corpus, CdmLogCode.ErrResolveReferenceFailure);

        // Keep for debugging
//        corpus.getCtx().getEvents().forEach(logEntry -> {
//            logEntry.forEach((key, value) -> System.out.println(key + "=" + value));
//            System.out.println();
//        });
    }

    /**
     * Tests events generated in StorageManager APIs
     */
    @Test
    public void testStorageManagerEvents() throws InterruptedException {
        CdmCorpusDefinition corpus = TestHelper.getLocalCorpus(TESTS_SUBPATH, "TestEventList");
        corpus.setEventCallback(eventCallback, CdmStatusLevel.Info, DUMMY_CORRELATION_ID);

        corpus.getStorage().mount("dummy", null);
        testBasicLogsState(corpus);
        TestHelper.assertCdmLogCodeEquality(corpus, CdmLogCode.ErrStorageNullAdapter);

        corpus.getStorage().unmount("nothing");
        testBasicLogsState(corpus);
        TestHelper.assertCdmLogCodeEquality(corpus, CdmLogCode.WarnStorageRemoveAdapterFailed);

        // No errors/warnings expected here
        corpus.getStorage().fetchRootFolder(null);
        testBasicLogsState(corpus);
        TestHelper.assertCdmLogCodeEquality(corpus, CdmLogCode.ErrStorageNullNamespace);

        corpus.getStorage().adapterPathToCorpusPath("Test");
        testBasicLogsState(corpus);
        TestHelper.assertCdmLogCodeEquality(corpus, CdmLogCode.ErrStorageInvalidAdapterPath);

        corpus.getStorage().corpusPathToAdapterPath("unknown:/Test");
        testBasicLogsState(corpus);
        TestHelper.assertCdmLogCodeEquality(corpus, CdmLogCode.ErrStorageAdapterNotFound);
        TestHelper.assertCdmLogCodeEquality(corpus, CdmLogCode.ErrStorageNamespaceNotRegistered);

        corpus.getStorage().createAbsoluteCorpusPath(null);
        testBasicLogsState(corpus);
        TestHelper.assertCdmLogCodeEquality(corpus, CdmLogCode.ErrPathNullObjectPath);

        corpus.getStorage().createRelativeCorpusPath(null);
        testBasicLogsState(corpus);
        TestHelper.assertCdmLogCodeEquality(corpus, CdmLogCode.ErrPathNullObjectPath);

    }

    /**
     * Test logging of API scope entry/exit.
     * @throws InterruptedException
     */
    @Test
    public void testScoping() throws InterruptedException {
        CdmCorpusDefinition corpus = TestHelper.getLocalCorpus(TESTS_SUBPATH, "TestEventList");
        // For scoping test we need to use log level Info
        corpus.setEventCallback(eventCallback, CdmStatusLevel.Info);
        corpus.getCtx().setCorrelationId(DUMMY_CORRELATION_ID);

        // Test fetching a good object, this should result event recorder empty
        corpus.fetchObjectAsync("local:/default.manifest.cdm.json").join();
        assertFalse(corpus.getCtx().getEvents().isRecording(), "Recording should be disabled at the end of API call");
        assertTrue(corpus.getCtx().getEvents().size() > 2,
                "There should have been at least 2 (debug) events recorded when fetching object with correct path");

        testBasicLogsState(corpus);

        // Verify method on entry/exit events
        assertEquals(corpus.getCtx().getEvents().get(0).get("method"), "fetchObjectAsync",
                "The first recorded event message should have specified scope callingfunc of 'fetchObjectAsync'");
        assertEquals(corpus.getCtx().getEvents().get(corpus.getCtx().getEvents().size() - 1).get("method"), "fetchObjectAsync",
                "The last recorded event message should have specified scope callingfunc of 'fetchObjectAsync'");

//         // Keep for debugging
//        corpus.getCtx().getEvents().forEach(logEntry -> {
//            logEntry.forEach((key, value) -> System.out.println(key + "=" + value));
//            System.out.println();
//         });
    }

    /**
     * Helper function to test that recording is stopped and no logs are recorded.
     * @param corpus CdmCorpusDefinition
     */
    private void testNoLogsState(CdmCorpusDefinition corpus) {
        testBasicLogsState(corpus, true);
    }

    /**
     * Helper function to check for event list state and tests that there are multiple entries and log enter/exit events were logged.
     * @param corpus CdmCorpusDefinition
     */
    private void testBasicLogsState(CdmCorpusDefinition corpus) {
        testBasicLogsState(corpus, false);
    }

    /**
     * Helper function to check for event list state and presence of scope enter/leave logs.
     * @param corpus CdmCorpusDefinition
     * @param expectNoLogs If true, tests that recording is stopped and there are no logs in EventList. If false,
     *                     tests that there are multiple entries and log enter/exit events were logged.
     */
    private void testBasicLogsState(CdmCorpusDefinition corpus, boolean expectNoLogs) {
        assertNotNull(corpus.getCtx().getEvents(), "Events list should not be null");
        assertFalse(corpus.getCtx().getEvents().isRecording(), "Recording should be disabled at the end of API call");

        if (expectNoLogs) {
            assertEquals(corpus.getCtx().getEvents().size(), 0,
                    "There should have been no events recorded when fetching object with correct path");
        } else {
            assertTrue(corpus.getCtx().getEvents().size() > 0, "There should have been at least one event recorded");
            assertTrue(corpus.getCtx().getEvents().get(0).containsKey("timestamp"), "The recorded event should have had a timestamp key");
            assertEquals(corpus.getCtx().getEvents().get(0).get("correlationId"), DUMMY_CORRELATION_ID,
                    "The recorded event should have had a correlationId key with the dummy value");

            if (corpus.getCtx().getReportAtLevel() == CdmStatusLevel.Progress) {
                assertEquals(corpus.getCtx().getEvents().get(0).get("message"), "Entering scope",
                        "The first recorded event message should have specified that new scope was entered");
                assertTrue(corpus.getCtx().getEvents().get(corpus.getCtx().getEvents().size() - 1).get("message").startsWith("Leaving scope"),
                        "The last recorded event message should have specified that new scope was exited");
            }
        }
    }
}
